File: src/main/java/org/manapart/endersort/Endersort.java
package org.manapart.endersort;

import net.minecraft.item.Item;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.Mod;

@Mod(Endersort.MODID)
@Mod.EventBusSubscriber(modid = Endersort.MODID)
public class Endersort {

    public static final String MODID = "endersort";
    public static final EndersortBlock ENDERSORT_BLOCK = createBlock();
    public static final EndersortItem ENDERSORT_ITEM = createItem(ENDERSORT_BLOCK);
    public static final EnderExtenderBlock ENDER_EXTENDER_BLOCK = createExtenderBlock();
    public static final EnderExtenderItem ENDER_EXTENDER_ITEM = createExtenderItem(ENDER_EXTENDER_BLOCK);
    public static final TileEntityType<EndersortEntity> tileType = createEntityType(ENDERSORT_BLOCK);
    public static Item endersortIcon = createIcon();

    public Endersort() {
        MinecraftForge.EVENT_BUS.register(this);
    }

    private static Item createIcon() {
        Item icon = new Item(new Item.Properties());
        icon.setRegistryName(MODID + ":es_icon");
        return icon;
    }

    private static EndersortBlock createBlock() {
        EndersortBlock block = new EndersortBlock();
        block.setRegistryName(new ResourceLocation(MODID + ":endersort_block"));
        return block;
    }

    private static EndersortItem createItem(EndersortBlock block) {
        EndersortItem item = new EndersortItem(block);
        item.setRegistryName(new ResourceLocation(MODID + ":endersort_item"));
        return item;
    }

    private static EnderExtenderBlock createExtenderBlock() {
        EnderExtenderBlock block = new EnderExtenderBlock();
        block.setRegistryName(new ResourceLocation(MODID + ":ender_extender_block"));
        return block;
    }

    private static EnderExtenderItem createExtenderItem(EnderExtenderBlock block) {
        EnderExtenderItem item = new EnderExtenderItem(block);
        item.setRegistryName(new ResourceLocation(MODID + ":ender_extender_item"));
        return item;
    }

    private static TileEntityType<EndersortEntity> createEntityType(EndersortBlock block) {
        TileEntityType.Builder<EndersortEntity> builder = TileEntityType.Builder.of(EndersortEntity::new, ENDERSORT_BLOCK);
        TileEntityType<EndersortEntity> tileType = builder.build(null);
        tileType.setRegistryName(MODID, "endersort");
        return tileType;
    }

}

File: src/main/java/org/manapart/endersort/Endersort.kt
package org.manapart.endersort

import net.minecraft.world.item.CreativeModeTabs
import net.minecraft.world.item.Item
import net.minecraftforge.event.BuildCreativeModeTabContentsEvent
import net.minecraftforge.fml.common.Mod
import org.manapart.endersort.ModBlocks.ENDERSORT_BLOCK
import org.manapart.endersort.ModBlocks.ENDER_EXTENDER_BLOCK
import org.manapart.endersort.ModItems.ENDERSORT_ITEM
import org.manapart.endersort.ModItems.ENDER_EXTENDER_ITEM
import thedarkcolour.kotlinforforge.forge.MOD_BUS

const val MODID = "endersort"

@Mod(MODID)
object Endersort {

    init {
        ModBlocks.REGISTRY.register(MOD_BUS)
        ModItems.REGISTRY.register(MOD_BUS)
        ModEntities.REGISTRY.register(MOD_BUS)
        MOD_BUS.addListener(::buildContents)
    }


    val icon by lazy { Item(Item.Properties()) }

    private fun buildContents(event: BuildCreativeModeTabContentsEvent) {
        if (event.tabKey == CreativeModeTabs.REDSTONE_BLOCKS) {
            event.accept(ENDERSORT_ITEM)
            event.accept(ENDER_EXTENDER_ITEM)
            event.accept(ENDERSORT_BLOCK)
            event.accept(ENDER_EXTENDER_BLOCK)
        }
    }

}